// *****************************************************************************
//
// This file is part of a MASA library or program.
// Refer to the included end-user license agreement for restrictions.
//
// Copyright (c) 2010 MASA Group
//
// *****************************************************************************

package MsgsClientToMessenger;

import "Common.proto";

// Drawings
message MsgShapeCreationRequest {
    required Common.MsgShape shape = 1; 
}

message MsgShapeUpdateRequest {
    required Common.DrawingId           shape     = 1;
    optional Common.RgbColor            color     = 2;
    optional Common.MsgLocation         location  = 3;
}

message MsgShapeDestructionRequest {
    required Common.DrawingId id = 1;
}

message MsgMarkerCreationRequest {
    required Common.MsgMarker           marker        = 1;      
}

message MsgMarkerUpdateRequest {
    required Common.MarkerId           marker      = 1;
    optional string                    name        = 2;
    optional string                    optional_information	= 3;
    optional string                    description = 4;
    optional Common.MarkerId           parent      = 5;
}

message MsgMarkerDestructionRequest {
    required Common.MarkerId            marker      = 1;
    required bool                       delete_all  = 2;
}

// client objects:

message MsgClientObjectCreationRequest {
    optional string name                            = 1;
    repeated Common.ClientObjectProperty properties = 2;
    optional bool persistent                        = 3;    // toggle persistence for this object
}

message MsgClientObjectUpdateRequest {
    required Common.ClientObjectId object           = 1;
    optional string name                            = 2;
    repeated Common.ClientObjectProperty properties = 3;
}

message MsgClientObjectDestructionRequest {
    required Common.ClientObjectId object           = 1;
}

message MsgClientToMessenger {
    message Content {
        optional MsgShapeCreationRequest           shape_creation_request            = 1;
        optional MsgShapeDestructionRequest        shape_destruction_request         = 2;
        optional MsgShapeUpdateRequest             shape_update_request              = 3;
        optional MsgMarkerCreationRequest          marker_creation_request           = 4;
        optional MsgMarkerUpdateRequest            marker_update_request             = 5;
        optional MsgMarkerDestructionRequest       marker_destruction_request        = 6;
        optional MsgClientObjectCreationRequest    client_object_creation_request    = 7;
        optional MsgClientObjectUpdateRequest      client_object_update_request      = 8;
        optional MsgClientObjectDestructionRequest client_object_destruction_request = 9;
    }
    optional int32   context = 1;
    required Content message = 2;
}
